# Data Template for AWAKEN Benchmark

https://app.box.com/s/nf4x11ubp20a00qntbexp4ukcfgzsb61/file/1592692746824

In [1]:
import netCDF4 as nc
import numpy as np
from datetime import datetime, timedelta

import xarray as xr
from netCDF4 import Dataset

# Add any possible locations of amr-wind-frontend here
amrwindfedirs = ['/projects/wind_uq/lcheung/amrwind-frontend/',
                 '/ccs/proj/cfd162/lcheung/amrwind-frontend/']
import sys, os, shutil, io
for x in amrwindfedirs: sys.path.insert(1, x)

%matplotlib inline

import postproamrwindsample_xarray as ppsample
import numpy as np
import matplotlib as plt


# Values from lidar_##.nc

In [2]:


#ncfile='/gpfs/lcheung/AWAKEN/Benchmark1/Phase1/FarmRuns/BM1_FarmRunProd1/postprocessing.shortsummit/lidar_48000.nc'
ncfile='/gpfs/lcheung/AWAKEN/Benchmark1/Phase1/FarmRuns/BM1_FarmRunProd1.summit/post_processing/lidar_68000.nc'
#heights are 10m to 950m every 10 m - 95 total
amr_heights = []
for i in range(10, 960,10):
    amr_heights.append(i)

group = 'lidar_siteA1' #'surface_siteA1'
#itimevec = np.arange(20285)  #for 48000
itimevec = np.arange(19991) # for 68000

ds = ppsample.getLineXR(ncfile, itimevec[:], ['velocityx', 'velocityy', 'velocityz','temperature', 'tke'], groupname=group, gettimes=True, includeattr=True)

A1velx = np.array([g for k, g in ds['velocityx'].items()])
A1vely = np.array([g for k, g in ds['velocityy'].items()])
A1velz = np.array([g for k, g in ds['velocityz'].items()])
A1temp = np.array([g for k, g in ds['temperature'].items()])
A1tke = np.array([g for k, g in ds['tke'].items()])


In [3]:
#print(len(A1velx) )
#print(ds['times'][-1])
#print(len(A1velx[2]))
#print(A1velx[2][-1])
#print(amr_heights)
print('AMR Wind time starts at', ds['times'][0])
print('AMR Wind time ends at', ds['times'][-1])
print(len(ds['times']))

AMR Wind time starts at 25999.999999970896
AMR Wind time ends at 27998.999999941807
19991


In [4]:
group = 'lidar_siteA2' #'surface_siteA1'
#itimevec = np.arange(300)

ds = ppsample.getLineXR(ncfile, itimevec[:], ['velocityx', 'velocityy', 'velocityz','temperature', 'tke'], groupname=group, gettimes=True, includeattr=True)

A2velx = np.array([g for k, g in ds['velocityx'].items()])
A2vely = np.array([g for k, g in ds['velocityy'].items()])
A2velz = np.array([g for k, g in ds['velocityz'].items()])
A2temp = np.array([g for k, g in ds['temperature'].items()])
A2tke = np.array([g for k, g in ds['tke'].items()])

In [5]:
group = 'lidar_siteB' #'surface_siteA1'
#itimevec = np.arange(300)

ds = ppsample.getLineXR(ncfile, itimevec[:], ['velocityx', 'velocityy', 'velocityz','temperature', 'tke'], groupname=group, gettimes=True, includeattr=True)

Bvelx = np.array([g for k, g in ds['velocityx'].items()])
Bvely = np.array([g for k, g in ds['velocityy'].items()])
Bvelz = np.array([g for k, g in ds['velocityz'].items()])
Btemp = np.array([g for k, g in ds['temperature'].items()])
Btke = np.array([g for k, g in ds['tke'].items()])

In [6]:
group = 'lidar_siteE36' #'surface_siteA1'
#itimevec = np.arange(300)

ds = ppsample.getLineXR(ncfile, itimevec[:], ['velocityx', 'velocityy', 'velocityz','temperature', 'tke'], groupname=group, gettimes=True, includeattr=True)

E36velx = np.array([g for k, g in ds['velocityx'].items()])
E36vely = np.array([g for k, g in ds['velocityy'].items()])
E36velz = np.array([g for k, g in ds['velocityz'].items()])
E36temp = np.array([g for k, g in ds['temperature'].items()])
E36tke = np.array([g for k, g in ds['tke'].items()])

# Surface Values from surface_##.nc

In [7]:
#s_ncfile='/gpfs/lcheung/AWAKEN/Benchmark1/Phase1/FarmRuns/BM1_FarmRunProd1/postprocessing.shortsummit/surf_68000.nc'
s_ncfile='/gpfs/lcheung/AWAKEN/Benchmark1/Phase1/FarmRuns/BM1_FarmRunProd1.summit/post_processing/surf_68000.nc'
#heights are at 4 and 10m 
s_amr_heights = [4,10]

group = 'surface_siteA1'
#itimevec = np.arange(300)

ds = ppsample.getLineXR(s_ncfile, itimevec[:], ['velocityx', 'velocityy', 'velocityz','temperature', 'tke'], groupname=group, gettimes=True, includeattr=True)

SA1velx = np.array([g for k, g in ds['velocityx'].items()])
SA1vely = np.array([g for k, g in ds['velocityy'].items()])
SA1velz = np.array([g for k, g in ds['velocityz'].items()])
SA1temp = np.array([g for k, g in ds['temperature'].items()])
SA1tke = np.array([g for k, g in ds['tke'].items()])


In [8]:
group = 'surface_siteA2'
#itimevec = np.arange(300)

ds = ppsample.getLineXR(s_ncfile, itimevec[:], ['velocityx', 'velocityy', 'velocityz','temperature', 'tke'], groupname=group, gettimes=True, includeattr=True)

SA2velx = np.array([g for k, g in ds['velocityx'].items()])
SA2vely = np.array([g for k, g in ds['velocityy'].items()])
SA2velz = np.array([g for k, g in ds['velocityz'].items()])
SA2temp = np.array([g for k, g in ds['temperature'].items()])
SA2tke = np.array([g for k, g in ds['tke'].items()])

In [9]:
group = 'surface_siteA5'
#itimevec = np.arange(300)

ds = ppsample.getLineXR(s_ncfile, itimevec[:], ['velocityx', 'velocityy', 'velocityz','temperature', 'tke'], groupname=group, gettimes=True, includeattr=True)

SA5velx = np.array([g for k, g in ds['velocityx'].items()])
SA5vely = np.array([g for k, g in ds['velocityy'].items()])
SA5velz = np.array([g for k, g in ds['velocityz'].items()])
SA5temp = np.array([g for k, g in ds['temperature'].items()])
SA5tke = np.array([g for k, g in ds['tke'].items()])

# Set up output file

48000.nc goes from 24000 to 26028.4 (20285)
-2023-08-24 04:55:00 to 
2023-08-24 05:28:48.400000

68000.nc goes from 26000 to 27999 (19991)
-  2023-08-24 05:28:20  to   
   2023-08-24 06:01:39

In [14]:

#_____________________________________________________________________________
# INPUT TO CHANGE BASED ON TIME RESOLUTION
time_resolution = 0.1 # [s]
#hours = .33
sec = 1999.105   #1200 - rounding errors made this wrong  2028.505   #
#1200 sec
#_____________________________________________________________________________
# DO NOT CHANGE ANYTHING AFTER THIS
#_____________________________________________________________________________

# Define dimensions
#time_length = int(hours*60*60/time_resolution)
time_length = int(sec/time_resolution)
#print(sec/time_resolution)
#print(time_length)
height_lidar_length = 37
height_assist_length = len([0, 10, 21, 33, 46, 61, 77, 95, 114, 136, 159, 185, 214, 245, 280, 318, 359, 405, 456, 512, 573, 640, 714, 795, 885, 983, 1092, 1211, 1342, 1486, 1645, 1819, 2011, 2223, 2455, 2710, 2991, 3300, 3640, 4014])

# Define time values
#start_time = datetime(2023, 8, 24, 0, 0) #This is the give start time of the benchmark 24 hours
start_time = datetime(2023, 8, 23, 22, 15)  #This is the start time of the AMR-Wind simulation
print(start_time)
print(start_time+ timedelta(seconds=ds['times'][0])) #This is the first time step to be saved in the simulation
#time_values = [start_time + timedelta(seconds=time_resolution*i) for i in range(time_length)] 
#put in start time + our time step values instead
time_values = [start_time + timedelta(seconds=ds['times'][i]) for i in range(len(ds['times']))]
print(time_values[0])
print(time_values[-1])
print(ds['times'][0])
print(ds['times'][-1])

# Define height values
height_lidar_values = np.arange(40, 410, 10)
height_assist_values = np.array([0, 10, 21, 33, 46, 61, 77, 95, 114, 136, 159, 185, 214, 245, 280, 318, 359, 405, 456, 512, 573, 640, 714, 795, 885, 983, 1092, 1211, 1342, 1486, 1645, 1819, 2011, 2223, 2455, 2710, 2991, 3300, 3640, 4014])


2023-08-23 22:15:00
2023-08-24 05:28:20
2023-08-24 05:28:20
2023-08-24 06:01:39
25999.999999970896
27998.999999941807


In [15]:
print(height_assist_values[30])
#print(amr_heights)
#print(A1velx)

1645


In [16]:
#Interpolate to correct heights
A1velx_int = np. interp(height_lidar_values, amr_heights, A1velx[0], left=np.nan,right = np.nan)

In [17]:

# Create netCDF file
with nc.Dataset("68000_inflow.nc", "w") as f:
    # Create dimensions
    f.createDimension("time", time_length)
    f.createDimension("height_lidar", height_lidar_length)
    f.createDimension("height_assist", height_assist_length)

    # Create variables
    u_siteA1 = f.createVariable("u_siteA1", "f4", ("time", "height_lidar"))  #lidar
    u_siteA2 = f.createVariable("u_siteA2", "f4", ("time", "height_lidar"))  #surf,lidar
    u_siteB = f.createVariable("u_siteB", "f4", ("time", "height_lidar"))
    u_siteE36 = f.createVariable("u_siteE36", "f4", ("time", "height_lidar"))
    v_siteA1 = f.createVariable("v_siteA1", "f4", ("time", "height_lidar"))
    v_siteA2 = f.createVariable("v_siteA2", "f4", ("time", "height_lidar"))
    v_siteB = f.createVariable("v_siteB", "f4", ("time", "height_lidar"))
    v_siteE36 = f.createVariable("v_siteE36", "f4", ("time", "height_lidar"))
    WS_siteA1 = f.createVariable("WS_siteA1", "f4", ("time", "height_lidar"))
    WS_siteA2 = f.createVariable("WS_siteA2", "f4", ("time", "height_lidar"))
    WS_siteB = f.createVariable("WS_siteB", "f4", ("time", "height_lidar"))
    WS_siteE36 = f.createVariable("WS_siteE36", "f4", ("time", "height_lidar"))
    w_siteA1 = f.createVariable("w_siteA1", "f4", ("time", "height_lidar"))
    w_siteA2 = f.createVariable("w_siteA2", "f4", ("time", "height_lidar"))
    w_siteB = f.createVariable("w_siteB", "f4", ("time", "height_lidar"))
    w_siteE36 = f.createVariable("w_siteE36", "f4", ("time", "height_lidar"))
    WD_siteA1 = f.createVariable("WD_siteA1", "f4", ("time", "height_lidar"))
    WD_siteA2 = f.createVariable("WD_siteA2", "f4", ("time", "height_lidar"))
    WD_siteB = f.createVariable("WD_siteB", "f4", ("time", "height_lidar"))
    WD_siteE36 = f.createVariable("WD_siteE36", "f4", ("time", "height_lidar"))
    TI_siteA1 = f.createVariable("TI_siteA1", "f4", ("time", "height_lidar"))
    TI_siteA2 = f.createVariable("TI_siteA2", "f4", ("time", "height_lidar"))
    TI_siteB = f.createVariable("TI_siteB", "f4", ("time", "height_lidar"))
    TI_siteE36 = f.createVariable("TI_siteE36", "f4", ("time", "height_lidar"))
    TKE_siteA1 = f.createVariable("TKE_siteA1", "f4", ("time", "height_lidar"))
    TKE_siteA2 = f.createVariable("TKE_siteA2", "f4", ("time", "height_lidar"))
    TKE_siteB = f.createVariable("TKE_siteB", "f4", ("time", "height_lidar"))
    TKE_siteE36 = f.createVariable("TKE_siteE36", "f4", ("time", "height_lidar"))
    # Temperature profiles
    T_siteB = f.createVariable("T_siteB", "f4", ("time", "height_assist"))
    T_siteE36 = f.createVariable("T_siteE36", "f4", ("time", "height_assist"))
    # Near-surface variables
    u_sfc_siteA1 = f.createVariable("u_sfc_siteA1", "f4", ("time",))
    u_sfc_siteA2 = f.createVariable("u_sfc_siteA2", "f4", ("time",))
    u_sfc_siteA5 = f.createVariable("u_sfc_siteA5", "f4", ("time",))
    v_sfc_siteA1 = f.createVariable("v_sfc_siteA1", "f4", ("time",))
    v_sfc_siteA2 = f.createVariable("v_sfc_siteA2", "f4", ("time",))
    v_sfc_siteA5 = f.createVariable("v_sfc_siteA5", "f4", ("time",))
    w_sfc_siteA1 = f.createVariable("w_sfc_siteA1", "f4", ("time",))
    w_sfc_siteA2 = f.createVariable("w_sfc_siteA2", "f4", ("time",))
    w_sfc_siteA5 = f.createVariable("w_sfc_siteA5", "f4", ("time",))
    WS_sfc_siteA1 = f.createVariable("WS_sfc_siteA1", "f4", ("time",))
    WS_sfc_siteA2 = f.createVariable("WS_sfc_siteA2", "f4", ("time",))
    WS_sfc_siteA5 = f.createVariable("WS_sfc_siteA5", "f4", ("time",))
    WD_sfc_siteA1 = f.createVariable("WD_sfc_siteA1", "f4", ("time",))
    WD_sfc_siteA2 = f.createVariable("WD_sfc_siteA2", "f4", ("time",))
    WD_sfc_siteA5 = f.createVariable("WD_sfc_siteA5", "f4", ("time",))
    TI_sfc_siteA1 = f.createVariable("TI_sfc_siteA1", "f4", ("time",))
    TI_sfc_siteA2 = f.createVariable("TI_sfc_siteA2", "f4", ("time",))
    TI_sfc_siteA5 = f.createVariable("TI_sfc_siteA5", "f4", ("time",))
    TKE_sfc_siteA1 = f.createVariable("TKE_sfc_siteA1", "f4", ("time",))
    TKE_sfc_siteA2 = f.createVariable("TKE_sfc_siteA2", "f4", ("time",))
    TKE_sfc_siteA5 = f.createVariable("TKE_sfc_siteA5", "f4", ("time",))
    L_sfc_siteA1 = f.createVariable("L_sfc_siteA1", "f4", ("time",))
    L_sfc_siteA2 = f.createVariable("L_sfc_siteA2", "f4", ("time",))
    L_sfc_siteA5 = f.createVariable("L_sfc_siteA5", "f4", ("time",))

    # Assign NaN values to variables
    for var in f.variables.values():
        var[:] = np.nan

    # Write time values to the file
    time_var = f.createVariable("time", "i8", ("time",))
    time_var.units = "seconds since 2023-08-23 22:15:00"
    time_var.calendar = "gregorian"
    time_var[:] = nc.date2num(time_values, units=time_var.units, calendar=time_var.calendar)

    height_lidar_var = f.createVariable("height_lidar", "i8", ("height_lidar",))
    height_lidar_var[:] = height_lidar_values
    height_assist_var = f.createVariable("height_assist", "i8", ("height_assist",))
    height_assist_var[:] = height_assist_values
    
    # Add attributes
    f.setncattr("time", "UTC dates")
    f.setncattr("height_lidar_units", "m a.g.l.")
    f.setncattr("height_assist_units", "m a.g.l.")

    #Add values from AMR Wind Run
    #u_siteA1 = f.createVariable("u_siteA1", "f4", ("time", "height_lidar"))
    #u_siteA1 =  
    

print("NetCDF file created successfully. NaN Values")





NetCDF file created successfully. NaN Values


In [18]:
#Put in interpolated AMR-Wind Values
# Open the netCDF file
dss = xr.open_dataset("68000_inflow.nc")

# Display dataset information
#print(dss.data_vars)
#print(dss['u_siteA1'])
#print(len(dss['time']))
#print(len(dss['u_siteA1'][5]))
for i in range(len(dss['time'])):
    #Height Lidar Values
    dss['u_siteA1'][i] =  np. interp(height_lidar_values, amr_heights, A1velx[i], left=np.nan,right = np.nan)
    dss['u_siteA2'][i] =  np. interp(height_lidar_values, amr_heights, A2velx[i], left=np.nan,right = np.nan)
    dss['u_siteB'][i] =  np. interp(height_lidar_values, amr_heights, Bvelx[i], left=np.nan,right = np.nan)
    dss['u_siteE36'][i] =  np. interp(height_lidar_values, amr_heights, E36velx[i], left=np.nan,right = np.nan)

    dss['v_siteA1'][i] =  np. interp(height_lidar_values, amr_heights, A1vely[i], left=np.nan,right = np.nan)
    dss['v_siteA2'][i] =  np. interp(height_lidar_values, amr_heights, A2vely[i], left=np.nan,right = np.nan)
    dss['v_siteB'][i] =  np. interp(height_lidar_values, amr_heights, Bvely[i], left=np.nan,right = np.nan)
    dss['v_siteE36'][i] =  np. interp(height_lidar_values, amr_heights, E36vely[i], left=np.nan,right = np.nan)

    dss['w_siteA1'][i] =  np. interp(height_lidar_values, amr_heights, A1velz[i], left=np.nan,right = np.nan)
    dss['w_siteA2'][i] =  np. interp(height_lidar_values, amr_heights, A2velz[i], left=np.nan,right = np.nan)
    dss['w_siteB'][i] =  np. interp(height_lidar_values, amr_heights, Bvelz[i], left=np.nan,right = np.nan)
    dss['w_siteE36'][i] =  np. interp(height_lidar_values, amr_heights, E36velz[i], left=np.nan,right = np.nan)

    dss['TKE_siteA1'][i] =  np. interp(height_lidar_values, amr_heights, A1tke[i], left=np.nan,right = np.nan)
    dss['TKE_siteA2'][i] =  np. interp(height_lidar_values, amr_heights, A2tke[i], left=np.nan,right = np.nan)
    dss['TKE_siteB'][i] =  np. interp(height_lidar_values, amr_heights, Btke[i], left=np.nan,right = np.nan)
    dss['TKE_siteE36'][i] =  np. interp(height_lidar_values, amr_heights, E36tke[i], left=np.nan,right = np.nan)

    
    #Height Assest Values
    dss['T_siteB'][i] = np. interp(height_assist_values, amr_heights, Btemp[i], left=np.nan,right = np.nan)
    dss['T_siteE36'][i] = np. interp(height_assist_values, amr_heights, E36temp[i], left=np.nan,right = np.nan)

    #Surface values - AMR-Wind 4m
    dss['u_sfc_siteA1'][i] = SA1velx[i][0]
    dss['u_sfc_siteA2'][i] = SA2velx[i][0]
    dss['u_sfc_siteA5'][i] = SA5velx[i][0]

    dss['v_sfc_siteA1'][i] = SA1vely[i][0]
    dss['v_sfc_siteA2'][i] = SA2vely[i][0]
    dss['v_sfc_siteA5'][i] = SA5vely[i][0]

    dss['w_sfc_siteA1'][i] = SA1velz[i][0]
    dss['w_sfc_siteA2'][i] = SA2velz[i][0]
    dss['w_sfc_siteA5'][i] = SA5velz[i][0]

    dss['TKE_sfc_siteA1'][i] = SA1tke[i][0]
    dss['TKE_sfc_siteA2'][i] = SA2tke[i][0]
    dss['TKE_sfc_siteA5'][i] = SA5tke[i][0]
    

In [19]:
#Calculated wind speed magnitude and direction
for i in range(len(dss['time'])):
    dss['WS_siteA1'][i] =  np.sqrt(dss['u_siteA1'][i]**2 + dss['v_siteA1'][i]**2 + dss['w_siteA1'][i]**2 )
    dss['WS_siteA2'][i] =  np.sqrt(dss['u_siteA2'][i]**2 + dss['v_siteA2'][i]**2 + dss['w_siteA2'][i]**2 )
    dss['WS_siteB'][i] =  np.sqrt(dss['u_siteB'][i]**2 + dss['v_siteB'][i]**2 + dss['w_siteB'][i]**2 )
    dss['WS_siteE36'][i] =  np.sqrt(dss['u_siteE36'][i]**2 + dss['v_siteE36'][i]**2 + dss['w_siteE36'][i]**2 )

    dss['WD_siteA1'][i] =  180+np.rad2deg(np.arctan(dss['u_siteA1'][i]/dss['v_siteA1'][i]))
    dss['WD_siteA2'][i] =  180+np.rad2deg(np.arctan(dss['u_siteA2'][i]/dss['v_siteA2'][i]))
    dss['WD_siteB'][i]  =  180+np.rad2deg(np.arctan(dss['u_siteB'][i]/dss['v_siteB'][i]))
    dss['WD_siteE36'][i] =  180+np.rad2deg(np.arctan(dss['u_siteE36'][i]/dss['v_siteE36'][i]))

    dss['WS_sfc_siteA1'][i] = np.sqrt(dss['u_sfc_siteA1'][i]**2 + dss['v_sfc_siteA1'][i]**2 + dss['w_sfc_siteA1'][i]**2 )
    dss['WS_sfc_siteA2'][i] = np.sqrt(dss['u_sfc_siteA2'][i]**2 + dss['v_sfc_siteA2'][i]**2 + dss['w_sfc_siteA2'][i]**2 )
    dss['WS_sfc_siteA5'][i] = np.sqrt(dss['u_sfc_siteA5'][i]**2 + dss['v_sfc_siteA5'][i]**2 + dss['w_sfc_siteA5'][i]**2 )

    dss['WD_sfc_siteA1'][i] =  180+np.rad2deg(np.arctan(dss['u_sfc_siteA1'][i]/dss['v_sfc_siteA1'][i]))
    dss['WD_sfc_siteA2'][i] =  180+np.rad2deg(np.arctan(dss['u_sfc_siteA2'][i]/dss['v_sfc_siteA2'][i]))
    dss['WD_sfc_siteA5'][i] =  180+np.rad2deg(np.arctan(dss['u_sfc_siteA5'][i]/dss['v_sfc_siteA5'][i]))

In [20]:
print(dss['T_siteB'][0].values )
print(height_lidar_values)

[      nan 302.982   303.64407 304.39993 305.105   305.9482  306.7315
 307.47662 308.159   308.80148 309.22342 309.6756  310.09073 310.4267
 310.71725 310.94672 311.11798 311.24777 311.34082 311.4135  312.48923
 321.0256  321.49866 321.5     321.5           nan       nan       nan
       nan       nan       nan       nan       nan       nan       nan
       nan       nan       nan       nan       nan]
[ 40  50  60  70  80  90 100 110 120 130 140 150 160 170 180 190 200 210
 220 230 240 250 260 270 280 290 300 310 320 330 340 350 360 370 380 390
 400]


In [21]:
#For testing
WS = np.sqrt(dss['u_siteA1'][0]**2 + dss['v_siteA1'][0]**2 + dss['w_siteA1'][0]**2 )
#WD2 = np.arctan2(dss['v_siteA1'][0][0].values,dss['u_siteA1'][0][0].values)
WD = np.arctan(dss['u_siteA1'][-1][5].values/dss['v_siteA1'][-1][5].values)
print(WD)
print(180+np.rad2deg(WD))
print(180+np.rad2deg(np.arctan(dss['u_siteA1'][-1][5].values/dss['v_siteA1'][-1][5].values)))
print(dss['WD_siteA1'][-1][5].values)
# print(np.sqrt(dss['u_siteA1'][0][1].values**2+ dss['v_siteA1'][0][1].values**2 + dss['w_siteA1'][0][1].values**2))
print(dss['u_siteA1'][0][0].values)
print(dss['v_siteA1'][0][0].values)
# print(dss['w_siteA1'][0].values)
#print(2.2125103/3.268545)
print(WS[5].values)
print(dss['WS_siteA1'][0][5].values)

-0.51466733
150.5117359161377
150.5117359161377
150.51173
-2.0071583
2.4236906
5.207794
5.207794


In [22]:
#write out dss into nc file
dss.close()  #Will overwrite
#dss.to_netcdf("SNL_ARM-Wind_Inflow.nc")

In [23]:
dss.to_netcdf("SNL_ARM-Wind_Inflow_68000.nc")